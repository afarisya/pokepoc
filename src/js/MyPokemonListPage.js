// JS
import React from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router'; // react-router v4
import {
    Container, 
    Row,
    Col,
    Card, 
    CardImg, 
    CardText, 
    CardBody,
    CardTitle, 
    CardSubtitle, 
    Modal, 
    ModalHeader, 
    ModalBody, 
    ModalFooter,
    Button,
    Input,
    Spinner
} from 'reactstrap';
import queryString from 'query-string'

import store from '../store';

// Components
import PokemonCard from './PokemonCard';
import PokemonPagination from './PokemonPagination';

// Actions
import { reqMyPokemonList, clearMyPokemons } from '../reducers/MyPokemonListReducers';


export class MyPokemonListPage extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
            searchQuery: "",
        }
	}

    componentWillMount() {
        this.setState({
            searchQuery: this.props.history.location.search
        }, () => {
            this.renderRoute();
        });
    }

    componentWillReceiveProps(nextProps) {
        if ( this.state.searchQuery !== nextProps.history.location.search ) {
            this.setState({
                searchQuery: nextProps.history.location.search
            }, () => {
                this.renderRoute();
            });
		}
	}

    componentWillUnmount() {
        store.dispatch(clearMyPokemons());
    }

    renderRoute = () => {
        var pageNumber = 1;

        const searchObj = queryString.parse(this.state.searchQuery);

        if ( typeof searchObj.page !== 'undefined' ) {
            pageNumber = parseInt(searchObj.page);
        }

        store.dispatch(reqMyPokemonList(pageNumber));
    }


	handlePageChange = (pageNumber) => { 
        const searchObj = queryString.parse(this.props.history.location.search);
        searchObj["page"] = pageNumber;
        var searchString = queryString.stringify(searchObj);
        this.props.history.push(this.props.history.location.pathname + "?" + searchString);
    }
    
    render() {
        return (
            <div>
                <Container>
                    { this.props.status !== "rcvMyPokemonList" ?
                        <Row className="pokemon-list-main-row">
                            <Spinner color="danger" style={{margin: "50% auto"}} />
                        </Row>
                        :
                        ( this.props.totalPokemons === 0 ?
                            <Row className="pokemon-list-main-row">
                                <Col xs="12">You haven't catch any pokemon</Col>
                            </Row>
                            :
                            <React.Fragment>
                                <div style={{padding: "30px 15px 0"}}>
                                    <h5>My Pokemons (Total {this.props.totalPokemons})</h5>
                                </div>
                                <Row className="pokemon-list-main-row">
                                    {
                                        this.props.pokemons.map((pokemon, index) => {
                                            return (
                                                <Col key={index} className="pokemon-col" xs="6" sm="4" md="3" lg="2">
                                                    <PokemonCard 
                                                        id={pokemon.pokemonId} 
                                                        name={pokemon.pokemonName} 
                                                        catched={true}
                                                        nickname={pokemon.pokemonNickname} 
                                                    />
                                                </Col>
                                            )
                                        })
                                    }
                                </Row>
                            </React.Fragment>
                        )
                    }
                    { this.props.totalPokemons !== 0 ?
                        <Row style={{padding: "0 13px"}}>
                            <PokemonPagination 
                                activePage={this.props.activePage}
                                totalItems={this.props.totalPokemons}
                                itemsPerPage={this.props.limit}
                                onChange={this.handlePageChange}
                            />
                        </Row>
                        : null
                    }
                </Container>
            </div>
        );
    }
};

function mapStateToProps(state) {
	return {
		status          : state.MyPokemonListReducers.status,
		activePage      : state.MyPokemonListReducers.activePage,
		offset          : state.MyPokemonListReducers.offset,
		limit           : state.MyPokemonListReducers.limit,
		totalPokemons   : state.MyPokemonListReducers.totalPokemons,
		pokemons        : state.MyPokemonListReducers.pokemons,
	}
}

export default withRouter(connect(mapStateToProps)(MyPokemonListPage));