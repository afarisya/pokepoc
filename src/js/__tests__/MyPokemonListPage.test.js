// MyPokemonListPage.test.js

import React from 'react';
import {
    makeMountRender,
    mockData,
    reduxify,
    snapshotify,
    ticks
} from '../../utils/test-utils';

import MyPokemonListPage from '../MyPokemonListPage';


describe("MyPokemonListPage status reqMyPokemonList. totalPokemons 0", () => {    
    const offset = 0;
    const limit  = 12;

    const wrapper = makeMountRender(
        reduxify(
            MyPokemonListPage,
            {},
            {
                MyPokemonListReducers: {
                    status: "reqMyPokemonList",
                    activePage: 1,
                    offset: offset,
                    limit: limit,
                    totalPokemons: 0,
                    pokemons: []
                }
            }
        )
    )();

    it('should matches snapshot', function() {
        expect(snapshotify(wrapper)).toMatchSnapshot();
    });

    it("should show spinner", () => { 
        expect(wrapper.find('.spinner-border').length).toEqual(1);
    });

    // it("should show pagination", () => {  
    //     expect(wrapper.find('.pagination').length).toEqual(1);
    // });
});

describe("MyPokemonListPage status reqMyPokemonList. totalPokemons 1", () => {    
    const offset = 0;
    const limit  = 12;

    const wrapper = makeMountRender(
        reduxify(
            MyPokemonListPage,
            {},
            {
                MyPokemonListReducers: {
                    status: "reqMyPokemonList",
                    activePage: 1,
                    offset: offset,
                    limit: limit,
                    totalPokemons: 1,
                    pokemons: [{pokemonId: "bulbasaur", pokemonName: "bulbasaur", pokemonNickname: "bulb"}]
                }
            }
        )
    )();

    it('should matches snapshot', function() {
        expect(snapshotify(wrapper)).toMatchSnapshot();
    });

    it("should show spinner", () => { 
        expect(wrapper.find('.spinner-border').length).toEqual(1);
    });

    it("should show pagination", () => {  
        expect(wrapper.find('.pagination').length).toEqual(1);
    });
});

describe("MyPokemonListPage status rcvMyPokemonList. totalPokemons 0", () => {    
    const offset = 0;
    const limit  = 12;

    const wrapper = makeMountRender(
        reduxify(
            MyPokemonListPage,
            {},
            {
                MyPokemonListReducers: {
                    status: "rcvMyPokemonList",
                    activePage: 1,
                    offset: offset,
                    limit: limit,
                    totalPokemons: 0,
                    pokemons: []
                }
            }
        )
    )();

    it('should matches snapshot', function() {
        expect(snapshotify(wrapper)).toMatchSnapshot();
    });

    it("should show 'You haven't catch any pokemon' if status = rcvMyPokemonList and pokemons.length = 0", () => {   
        expect(wrapper.find('#no-pokemons').length).toBeGreaterThan(0);
    });
});

describe("MyPokemonListPage status rcvMyPokemonList. totalPokemons 1", () => {    
    const offset = 0;
    const limit  = 12;

    const wrapper = makeMountRender(
        reduxify(
            MyPokemonListPage,
            {},
            {
                MyPokemonListReducers: {
                    status: "rcvMyPokemonList",
                    activePage: 1,
                    offset: offset,
                    limit: limit,
                    totalPokemons: 1,
                    pokemons: [{pokemonId: "bulbasaur", pokemonName: "bulbasaur", pokemonNickname: "bulb"}]
                }
            }
        )
    )();

    it('should matches snapshot', function() {
        expect(snapshotify(wrapper)).toMatchSnapshot();
    });

    it('show pokemon bulbasaur in list', function() {
        expect(wrapper.find('.pokemon-nickname').text()).toEqual('bulb');
        expect(wrapper.find('a .badge').text()).toEqual('bulbasaur');
    });
    
    it('can click on pokemon bulbasaur', function() {
        wrapper.find('a').first().simulate('click');
        // expect(window.location.href).toEqual(expect.any(String));        
    });

    it("should show pagination", () => {  
        expect(wrapper.find('.pagination').length).toEqual(1);
    });
});