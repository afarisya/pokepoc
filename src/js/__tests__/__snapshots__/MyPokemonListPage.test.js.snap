// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPokemonListPage should render my pokemon list page if status = rcvMyPokemonList 1`] = `
Array [
  <div>
    <Container
      tag="div"
    >
      <Row
        className="pokemon-list-main-row"
        tag="div"
      >
        <Col
          id="no-pokemons"
          tag="div"
          widths={
            Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ]
          }
          xs="12"
        >
          You haven't catch any pokemon
        </Col>
      </Row>
    </Container>
  </div>,
]
`;

exports[`MyPokemonListPage should render my pokemon list page if status = reqMyPokemonList 1`] = `
Array [
  <div>
    <Container
      tag="div"
    >
      <Row
        className="pokemon-list-main-row"
        tag="div"
      >
        <Spinner
          color="danger"
          style={
            Object {
              "margin": "50% auto",
            }
          }
          tag="div"
          type="border"
        >
          Loading...
        </Spinner>
      </Row>
    </Container>
  </div>,
]
`;

exports[`MyPokemonListPage status rcvMyPokemonList. totalPokemons 0 should matches snapshot 1`] = `"<div><div class=\\"container\\"><div class=\\"pokemon-list-main-row row\\"><div id=\\"no-pokemons\\" class=\\"col-12\\">You haven't catch any pokemon</div></div></div></div>"`;

exports[`MyPokemonListPage status rcvMyPokemonList. totalPokemons 1 should matches snapshot 1`] = `"<div><div class=\\"container\\"><div style=\\"padding: 30px 15px 0px;\\"><h5>My Pokemons (Total 1)</h5></div><div class=\\"pokemon-list-main-row row\\"><div class=\\"pokemon-col col-6 col-sm-4 col-md-3 col-lg-2\\"><a href=\\"/pokemon/bulbasaur\\"><div class=\\"card\\"><img width=\\"100%\\" src=\\"\\" alt=\\"Card image cap\\" class=\\"card-img-top\\"><div class=\\"card-body\\"><div class=\\"card-title\\"><div class=\\"pokemon-nickname\\">bulb</div><span class=\\"badge badge-light badge-pill\\">bulbasaur</span></div><button type=\\"button\\" class=\\"release-btn btn btn-warning\\">Release</button></div></div></a></div></div><div style=\\"padding: 0px 13px;\\" class=\\"row\\"><div style=\\"margin-bottom: 20px;\\"><nav class=\\"\\" aria-label=\\"Page navigation example\\"><ul class=\\"pagination\\"><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"First\\"><span aria-hidden=\\"true\\">«</span><span class=\\"sr-only\\">First</span></button></li><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"Previous\\"><span aria-hidden=\\"true\\">‹</span><span class=\\"sr-only\\">Previous</span></button></li><li class=\\"page-item active\\"><button class=\\"page-link\\">1</button></li><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"Next\\"><span aria-hidden=\\"true\\">›</span><span class=\\"sr-only\\">Next</span></button></li><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"Last\\"><span aria-hidden=\\"true\\">»</span><span class=\\"sr-only\\">Last</span></button></li></ul></nav></div></div></div></div>"`;

exports[`MyPokemonListPage status reqMyPokemonList. totalPokemons 0 should matches snapshot 1`] = `"<div><div class=\\"container\\"><div class=\\"pokemon-list-main-row row\\"><div role=\\"status\\" style=\\"margin: 50% auto;\\" class=\\"spinner-border text-danger\\"><span class=\\"sr-only\\">Loading...</span></div></div></div></div>"`;

exports[`MyPokemonListPage status reqMyPokemonList. totalPokemons 1 should matches snapshot 1`] = `"<div><div class=\\"container\\"><div class=\\"pokemon-list-main-row row\\"><div role=\\"status\\" style=\\"margin: 50% auto;\\" class=\\"spinner-border text-danger\\"><span class=\\"sr-only\\">Loading...</span></div></div><div style=\\"padding: 0px 13px;\\" class=\\"row\\"><div style=\\"margin-bottom: 20px;\\"><nav class=\\"\\" aria-label=\\"Page navigation example\\"><ul class=\\"pagination\\"><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"First\\"><span aria-hidden=\\"true\\">«</span><span class=\\"sr-only\\">First</span></button></li><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"Previous\\"><span aria-hidden=\\"true\\">‹</span><span class=\\"sr-only\\">Previous</span></button></li><li class=\\"page-item active\\"><button class=\\"page-link\\">1</button></li><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"Next\\"><span aria-hidden=\\"true\\">›</span><span class=\\"sr-only\\">Next</span></button></li><li class=\\"page-item disabled\\"><button class=\\"page-link\\" aria-label=\\"Last\\"><span aria-hidden=\\"true\\">»</span><span class=\\"sr-only\\">Last</span></button></li></ul></nav></div></div></div></div>"`;
